KEYCLOAK-7675 Support for Device Authorization Grant

Michito-Okai opened the pull request:
https://github.com/keycloak/keycloak/pull/6992
Because the above PR was stopped, I tried to support for Device Authorization Grant by doing the following.
-	Resolve conflicts
-	Add a function to set “device code lifespan” and “polling interval” for each client
-	Fix the behave when consent is canceled by the end user on the consent screen
-	Fix integration tests

JIRA ticket
https://issues.redhat.com/browse/KEYCLOAK-7675

I would like to get your feedback.

wadahiro committed:
KEYCLOAK-7675 Prototype Implementation of Device Authorization Grant.

Author:    Hiroyuki Wada <h2-wada@nri.co.jp>
Date:      Thu May 2 00:22:24 2019 +0900

Signed-off-by: Łukasz Dywicki <luke@code-house.org>

wadahiro committed:
KEYCLOAK-7675 SPI and default implementation for Device User Code.

Author:    Hiroyuki Wada <h2-wada@nri.co.jp>
Date:      Sun May 12 15:47:15 2019 +0900

Signed-off-by: Łukasz Dywicki <luke@code-house.org>

splatch committed:
Fix failing test, cause oauth device flow is encoded using realm attributes.

splatch committed:
Make sure additional params are passed between device request and user authnetication.

Michito-Okai committed:
KEYCLOAK-7675 Support for Device Authorization Grant

pedroigor commented:
@Michito-Okai Thanks for pushing this forward. We do want to have support for device flow, but right now we are reviewing CIBA changes.

As it was prioritized, discussions in the PR are already happening and we should expect it merged soon. Some of the capabilities we would be able to leverage from CIBA changes are mainly related to pooling and code re-use. As well as the overall design considering they pretty much similar in terms of functionality.

pedroigor reviewed (changes requested):
@Michito-Okai Are you able to check this branch https://github.com/pedroigor/keycloak/tree/KEYCLOAK-7675?

I've added a commit with some minor changes:

* Tests
* Discovery document
* Rebase + Fixes

Otherwise, I think these changes are good enough. We should also have docs and I can work on them.

Michito-Okai commented:
@pedroigor  
I incorporated your minor changes into this pull request. Please confirm this.

pedroigor commented:
@Michito-Okai Changes are here. The new test is failing, going to check what is going on.

pedroigor commented:
@Michito-Okai Tests are OK. Not sure why they are failing.

I did some changes here https://github.com/pedroigor/keycloak/tree/KEYCLOAK-7675. Please, look at the last commit.

Basically, I'm trying to simplify the code in regard to device configuration. This is similar to what I'm thinking for CIBA where we need to start centralizing changes related to extensions in a single place so that in the future we can refactor a bit how we add these extensions more easily.

Please, let me know your thoughts and if you want to include the changes here.

I think that is the last thing from my side. We should be able to merge soon.

Michito-Okai commented:
@pedroigor 
> Please, let me know your thoughts and if you want to include my changes.

I agree with your opinion.

> * Create a package on `keycloak-services` specific for device flow
> * Put in this package the `OAuth2DeviceAuthorizationEndpoint`
> * Extract code from `TokenEndpoint` related to device flow and create a new `DeviceGrantType` and put this class in the same package

I would like to make the above improvements.

Michito-Okai commented:
@pedroigor 
> * Create a package on `keycloak-services` specific for device flow
> * Put in this package the `OAuth2DeviceAuthorizationEndpoint`
> * Extract code from `TokenEndpoint` related to device flow and create a new `DeviceGrantType` and put this class in the same

I incorporated your minor changes and the above improvements into this pull request. Also, since the client attribute names of `device code lifespan` and `polling interval` were not defined in OAuth2DeviceConfig.java, `device code lifespan` and `polling interval` set on the client setting screen were not reflected. So I modified `OAuth2DeviceConfig.java` and `OAuth2DeviceAuthorizationGrantTest.java` to reflect `device code lifespan` and `polling interval` set on the client setting screen. Please confirm this.

pedroigor commented:
> @pedroigor
> 
> > * Create a package on `keycloak-services` specific for device flow
> > * Put in this package the `OAuth2DeviceAuthorizationEndpoint`
> > * Extract code from `TokenEndpoint` related to device flow and create a new `DeviceGrantType` and put this class in the same
> 
> I incorporated your minor changes and the above improvements into this pull request. Also, since the client attribute names of `device code lifespan` and `polling interval` were not defined in OAuth2DeviceConfig.java, `device code lifespan` and `polling interval` set on the client setting screen were not reflected. So I modified `OAuth2DeviceConfig.java` and `OAuth2DeviceAuthorizationGrantTest.java` to reflect `device code lifespan` and `polling interval` set on the client setting screen. Please confirm this.

@Michito-Okai I have looked at the changes and I should be sending more changes soon.

I also looking that the user interaction when authorizing the device. We need to authenticate the user prior to validating the code, right?

Currently, it is doing the opposite. Different than specs and in a way that does not make much sense, as anyone in possession of the code could approve the device.

Michito-Okai commented:
@pedroigor 
> I also looking that the user interaction when authorizing the device. We need to authenticate the user prior to validating the code, right?

[Section-3.3 of RFC8628](https://tools.ietf.org/html/rfc8628#section-3.3) states `Authorization servers supporting this specification MUST implement a user-interaction sequence that starts with the user navigating to "verification_uri" and continues with them supplying the "user_code" at some stage during the interaction.  Other than that, the exact sequence and implementation of the user interaction is up to the authorization server;`. In fact, [Keycloak community](https://github.com/keycloak/keycloak-community/blob/master/design/oauth2-device-authorization-grant.md) has decided to support User Code Verification First.

pedroigor commented:
@Michito-Okai Could you please check my latest changes https://github.com/pedroigor/keycloak/tree/KEYCLOAK-7675 ?

Basically:

* Moving more code to the device package
* Reducing changes to core classes
* Minor changes to test scenarios

I might have changed how events are fired and I removed them from tests, for now.

Sorry for another round of changes. But I think now we are 95% close to the end of my review.

Something I'm still not sure about is:

* Right now we are asking the user for code and then authenticating. Ideally, we should do the verification only after authenticating the user. This should be possible when doing `verification_uri_complete` together with a required action. But when the flow is started with no user code, then it is hard to figure out the client to which we need to bind the authentication session (when redirecting first to authenticate).

At the same time, I think this approach is fine after thinking more about it. We don't actually do anything with the code but just some basic validation. The code is also bound to the user so only he/she would be able to approve once authenticated.

Please, let me know your thoughts and if my changes make sense.

pedroigor commented:
> [Section-3.3 of RFC8628](https://tools.ietf.org/html/rfc8628#section-3.3) states `Authorization servers supporting this specification MUST implement a user-interaction sequence that starts with the user navigating to "verification_uri" and continues with them supplying the "user_code" at some stage during the interaction. Other than that, the exact sequence and implementation of the user interaction is up to the authorization server;`. In fact, [Keycloak community](https://github.com/keycloak/keycloak-community/blob/master/design/oauth2-device-authorization-grant.md) has decided to support User Code Verification First.

OK, then. Thanks for the heads up. I tried something with a required action but then we need to bind the authentication session to the client associated with the user code during the authentication process, which is tricky.

pedroigor commented:
@Michito-Okai Hey. Did you have time to check my changes? I think we can have this PR merged this week. Pretty much done from my side.

Michito-Okai commented:
@pedroigor 
I'm checking now. I plan to commit in a few hours.

Michito-Okai commented:
Sorry for the late commit.
I incorporated your more changes and fixed the following two points.
- Set Device Authorization Endpoint correctly in OIDCWellKnown.
- Add processing to update client session and user session in `oauth2DeviceFlow()` of `DeviceGrantType.java`
  Those processing are necessary, but I accidentally commented them out at my last commit.

I have one question about your more changes. Device Authorization Endpoint changes from `/realms/{realmName}/openid-connect/device/auth` to `/realms/{realmName}/openid-connect/auth/device/auth` by Incorporating your changes. [Keycloak community](https://github.com/keycloak/keycloak-community/blob/master/design/oauth2-device-authorization-grant.md) has decided to define Device Authorization Endpoint as `/realms/{realmName}/openid-connect/device/auth`. Is this change in line with your idea?

pedroigor commented:
> Sorry for the late commit.
> I incorporated your more changes and fixed the following two points.
> 
> * Set Device Authorization Endpoint correctly in OIDCWellKnown.
> * Add processing to update client session and user session in `oauth2DeviceFlow()` of `DeviceGrantType.java`
>   Those processing are necessary, but I accidentally commented them out at my last commit.

Do you mean this?

```
tokenEndpoint.updateClientSession(clientSession);
tokenEndpoint.updateUserSessionFromClientAuth(userSession);
```

Btw, regarding updating user session with 

The first invocation is related to `client_session_state` and Keycloak adapters which is not really necessary for device flow?

W.r.t. to the second invocation. I'm not sure where it is being used. 

That is why I thought that both were OK to remove ...

> 
> I have one question about your more changes. Device Authorization Endpoint changes from `/realms/{realmName}/openid-connect/device/auth` to `/realms/{realmName}/openid-connect/auth/device/auth` by Incorporating your changes. [Keycloak community](https://github.com/keycloak/keycloak-community/blob/master/design/oauth2-device-authorization-grant.md) has decided to define Device Authorization Endpoint as `/realms/{realmName}/openid-connect/device/auth`. Is this change in line with your idea?

I did that because we should be dropping the `/auth` in favor of `/` as the root path. That is what we are already doing for Quarkus. It might look strange now, but it should look better :) I'm going to confirm this with others in the team.

pedroigor commented:
@Michito-Okai One thing that I was wondering is if we should use tokens using transient sessions for device flow. 

Spec-wise, the refresh token is optional. Depending on the use case, we could avoid creating user sessions (and avoid keeping its state in the cluster). I'm not pushing for having this now, but I'm wondering if there are use cases, in your opinion, that don't require a refresh token at all. I can think of some.

kfox1111 commented:
I think k8s cli clients on a shared cluster (like hpc) login nodes may be a good use case for not offering refresh tokens.

You can log into the client then for a little bit, and then need to login again later instead of the session being able to outlast your ssh session.

pedroigor commented:
> I think k8s cli clients on a shared cluster (like hpc) login nodes may be a good use case for not offering refresh tokens.
> 
> You can log into the client then for a little bit, and then need to login again later instead of the session being able to outlast your ssh session.

For client credentials, we already support transient sessions where refresh tokens are no longer issued. We also provide the option, for backward compatibility purposes (but planned to be removed), to return RTs.

Michito-Okai commented:
I am making the following changes now.
1. remove processings to update client session and user session in `oauth2DeviceFlow()` of `DeviceGrantType.java`

> ```
> tokenEndpoint.updateClientSession(clientSession);
> tokenEndpoint.updateUserSessionFromClientAuth(userSession);
> ```
> 
> The first invocation is related to `client_session_state` and Keycloak adapters which is not really necessary for device flow?
> 
> W.r.t. to the second invocation. I'm not sure where it is being used.
> 
> That is why I thought that both were OK to remove ...

I'm sorry. Certainly, the above processings were unnecessary for the device flow.

2. change endpoints
> My answer to the last question does not make sense. Yeah, we can revert to what was agreed in the design document. That was an attempt to have the device authorization endpoint within the existing `/auth` path. So it would be `/auth/device` and not `/auth/device/auth`. But I don't care about changing this now if you think we should not be doing it.

If Device Authorization Endpoint changes along the way, Keycloak users also need to change Device Authorization Endpoint. So considering operability, I think it's better to set it to `/realms/{realmName}/openid-connect/auth/device` now. This time, I change the endpoint as follows.

Device Authorization Endpoint
before: `/realms/{realm}/protocol/openid-connect/auth/device/auth`
after : `/realms/{realm}/protocol/openid-connect/auth/device`

Verification Endpoint
before: `/realms/{realm}/device`
after : `/realms/{realm}/device`

User Code Verification Endpoint
before: `/realms/{realm}/login-actions/device`
after : `/realms/{realm}/login-actions/device/user-code`

Michito-Okai commented:
@pedroigor 
> Spec-wise, the refresh token is optional. Depending on the use case, we could avoid creating user sessions (and avoid keeping its state in the cluster). I'm not pushing for having this now, but I'm wondering if there are use cases, in your opinion, that don't require a refresh token at all. I can think of some.

I think that the fact that refresh tokens are optional means that some users want to use refresh tokens. Therefore, I think it is better to issue a refresh token.

mposolda commented on server-spi/src/main/java/org/keycloak/models/ClientModel.java:
```diff
@@ -199,6 +200,15 @@
     boolean isServiceAccountsEnabled();
     void setServiceAccountsEnabled(boolean serviceAccountsEnabled);
 
+    default boolean isOAuth2DeviceAuthorizationGrantEnabled() {
```
I think it will be better to have this inside the OIDCAdvancedConfigWrapper class instead of ClientModel class. OIDCAdvancedConfigWrapper already contains all of OIDC related properties.

Michito-Okai replied:
I fixed.

mposolda replied:
I see there is still constant OAUTH2_DEVICE_AUTHORIZATION_GRANT_ENABLED on the ClientModel. Can this be removed from ClientModel?

Actually, I figured that there is separate class OAuth2DeviceConfig added for all the properties related to device-grant. So probably it will be even better to remove the configuration from OIDCAdvancedConfigWrapper and move the configuration option into the OAuth2DeviceConfig class. That will be the cleanest approach IMO. WDYT?

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,488 @@
+        // https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.1
+        // The spec says "The client authentication requirements of Section 3.2.1 of [RFC6749]
+        // apply to requests on this endpoint".
+        AuthorizeClientUtil.ClientAuthResult clientAuth = AuthorizeClientUtil.authorizeClient(session, event);
```
I think it will be needed to rebase this PR. In latest master, signature of AuthorizeClientUtil.authorizeClient slightly changed.

Michito-Okai replied:
I fixed.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,488 @@
+            ServicesLogger.LOGGER.oidcScopeMissing();
+        }
+
+        createAuthenticationSession(client, request);
```
I am not sure what exactly is the reason of create authenticationSession here. Is it really needed? It looks that endpoint to deviceUrl needs just to create deviceCode and userCode and save the stuff to infinispan actionTokens cache. But authenticationSession is probably not needed? I think it is sufficient to create it later once user opens the URL of the browser?

pedroigor replied:
I agree, not needed.

Michito-Okai replied:
I fixed.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,488 @@
+        AuthenticationSessionModel authenticationSession = null;
+
+        // try to recover the authentication session in case the user is pushing the code
+        if (HttpMethod.POST.equalsIgnoreCase(httpRequest.getHttpMethod())) {
```
Instead of checking HttpMethod here, it will be slightly better if there is private method introduced like `verifyUserCode(String userCode, boolean isActionRequest)` .  When this private method is called from the GET method, it will use `isActionRequest` as false. When there is a call from the verifyUserCode POST method, it will use true. This will also allow to avoid duplicating creation of authenticationSession as I described below.

Michito-Okai replied:
I fixed.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,488 @@
+
+            ClientModel client = realm.getClientByClientId(deviceCode.getClientId());
+
+            authenticationSession = createAuthenticationSession(client);
```
This is wrong as it will sometimes creates the duplicated authenticationSession. For the case when you don't use `verification_uri_complete` and the screen is confirmed, this method is called from the `verifyUserCode` POST method. That checks the existing authenticationSession (lines 184-197), but then also creates the new authenticationSession here.

If it is done the approach with `isActionRequest` parameter, you will create authenticationSession here just if this is not actionRequest. For the actionRequest, there should be a hard requirement that authenticationSession already exists (as done in the block 184-197).

pedroigor replied:
Actually, we should avoid creating authentication sessions when doing GET. But only when doing POST + valid `user_code`.

The verification page should be completely stateless where an authentication session is only needed when redirecting the user to authenticate and confirm the consent.

mposolda replied:
@pedroigor +1 . Considering that we will always show consent page, it is fine to postpone creation of authenticationSession to the latest possible stage.

Michito-Okai replied:
I fixed.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,488 @@
+        return client;
+    }
+
+    private ClientModel checkClient(String clientId, AuthenticationSessionModel authenticationSession) {
```
There is big amount of duplicated code needed to handle client verification. Isn't it possible to re-use some basic checks (EG. call to client authentication, checking if client enabled, if OIDC login protocol, if bearer-only etc)? For example `AuthorizationEndpoint.checkClient` already uses lots of same verifications . Will it be better if DeviceEndpoint inherits from AuthorizationEndpoint and calls `super.checkClient` with possibly some more checks added specific to device flow? Like for example call to `client.isOAuth2DeviceAuthorizationGrantEnabled()` , which (if I am not missing something) is the only check specific to deviceFlow.

pedroigor replied:
Makes sense. As we have similar duplication in other places, refactoring should probably consider them too.

I don't see a problem in re-using some validations from there, but I won't push for this in this PR too.

We are trying to keep device flow-related code as much as possible in a single place. The idea is to look at the changes here to find gaps and improve how people can create extensions without having to touch different parts and duplicate code. For instance, from this PR you have that:

* We need a GrantType SPI (there are some initial discussions around it)
* We need to provide an SPI for extending the protocol API
* We need to provide ways to extend/re-use code from authorization and token endpoints

My point being, that we should think about this more carefully instead of extending what was not carefully thought for being extended.

mposolda commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanOAuth2DeviceTokenStoreProvider.java:
```diff
@@ -0,0 +1,241 @@
+    public OAuth2DeviceCodeModel getByDeviceCode(RealmModel realm, String deviceCode) {
+        try {
+            BasicCache<String, ActionTokenValueEntity> cache = codeCache.get();
+            ActionTokenValueEntity existing = cache.get(OAuth2DeviceCodeModel.createKey(realm, deviceCode));
```
There are some side-effects of having long keys for infinispan caches.

I wonder that one possible improvement to this is, that the realm UUID is not added to the key of the infinispan entries. Instead of this, realmId would be added as the field in the entity classes and all the lookup methods will programatically check if the realmId matches with the "realm" given as an argument to this method. This applies probably to all entities used by this class (not just deviceEntity, but also to userEntity and "polling" entity).

Michito-Okai replied:
I fixed.

mposolda reviewed (changes requested):
@Michito-Okai @pedroigor Thanks a lot for this PR and great work to move this forward!

I've did the review of the code from yesterday and have few comments. Now I see that this PR is still being actively worked and updated based on the review from @pedroigor, so there is a chance that some of my review comments are not applicable anymore? However I believe that most of them are still applicable.

pedroigor commented:
> User Code Verification Endpoint
> before: `/realms/{realm}/login-actions/device`
> after : `/realms/{realm}/login-actions/device/user-code`

Sounds good to me, thanks. Just a comment about the above. Do we need it at `/login-actions` ? Or maybe we could have it in `/realms/{realm}/device`? My point being that this is basically the endpoint handling the submission of the user verification code after making a GET to `/realms/{realm}/device`, right?

And the same goes with the `status`?

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,488 @@
+            response.setVerificationUri(deviceUrl);
+            response.setVerificationUriComplete(deviceUrl + "?user_code=" + response.getUserCode());
+
+            return Response.ok(JsonSerialization.writeValueAsBytes(response)).type(MediaType.APPLICATION_JSON_TYPE).build();
```
I agree. We don't need CORS.

Michito-Okai replied:
I fixed.

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/DeviceGrantType.java:
```diff
@@ -0,0 +1,327 @@
+                Response.Status.BAD_REQUEST);
+        }
+
+        tokenEndpoint.updateClientSession(clientSession);
```
I agree, except for CORS support.

Michito-Okai replied:
I fixed.

mposolda commented:
@Michito-Okai As suggested by @pedroigor, I am trying to summary, how I see the ideal flow regarding the authenticationSession creation and loading. Most of the points are already covered in the inline comments and some of the points already works in the PR exactly as as I mention here, so no changes needed in the PR for those. So just a summary:

- AuthenticationSession doesn't need to be created at all in the initial deviceRequest (currently it is. At least in the PR version from yesterday)
- In case that `verification_uri_complete` is used, the authenticationSession should be created at the moment when the request is processed
- In case that `verification_uri_complete` is not used, the authenticationSession should be created at the moment of initial GET request. Once user sends the userCode, the POST request must verify that it is able to join the existing authenticationSession. It shouldn't create new authenticationSession (which it currently does and that is a bug IMO)
- For both cases with and without `verification_uri_complete`, there is redirect to LoginActionsService done, which should make sure that authenticationSession already exists.
- There is consent screen always displayed after the authentication, which is good IMO, and it matches the requirement from the 5.4 section of the specification [1] that user must always consent the device

[1] https://tools.ietf.org/html/rfc8628#section-5.4

mposolda commented:
@Michito-Okai @pedroigor BTV. I think that PR will require also some documentation though? At least some short notice worth to be added somewhere there https://www.keycloak.org/docs/latest/securing_apps/index.html#other-openid-connect-libraries . Probably into the "endpoints" section?

pedroigor commented:
@mposolda PR already marked with the label for docs.

Michito-Okai commented:
@pedroigor 
> Sounds good to me, thanks. Just a comment about the above. Do we need it at `/login-actions` ? Or maybe we could have it in `/realms/{realm}/device`? My point being that this is basically the endpoint handling the submission of the user verification code after making a GET to `/realms/{realm}/device`, right? We can also drop the `user-code` because that request is a POST, so it should be fine to have it without any sub-path. So it would be `/realms/{realm}/device` (same verification endpoint).
> 
> And the same goes with the `status`?

`DeviceEndpoint.java` has four endpoints. They are set below at the time of the current PR.
1. Device Authorization Endpoint
   method: POST
   path: /realms/{realm}/protocol/openid-connect/auth/device/auth
   @Path: "/auth"
2. Verification Endpoint
   method: GET
   path: /realms/{realm}/device
   @Path: ""
3. User Code Verification Endpoint
   method: POST
   path: /realms/{realm}/login-actions/device
   @Path: "/"
4. Status Endpoint
   method: GET
   path: /auth/realms/DeviceFlow/login-actions/device/status
   @Path: "status"


If we modified four endpoints as the quote above, the endpoint would be set:

1. Device Authorization Endpoint
   method: POST
   path: /realms/{realm}/protocol/openid-connect/auth/device
   @Path: ""
2. Verification Endpoint
   method: GET
   path: /realms/{realm}/device
   @Path: ""
3. User Code Verification Endpoint
   method: POST
   path: /realms/{realm}/device
   @Path: ""
4. Status Endpoint
   method: GET
   path: /realms/{realm}/device/status
   @Path: "status"

User Code Verification Endpoint's @path overlaps Device Authorization Endpoint's @path. Therefore I want to set them as the following.

1. Device Authorization Endpoint
   method: POST
   path: /realms/{realm}/protocol/openid-connect/auth/device
   @Path: ""
2. Verification Endpoint
   method: GET
   path: /realms/{realm}/device
   @Path: ""
3. User Code Verification Endpoint
   method: POST
   path: /realms/{realm}/device/user-code
   @Path: "/user-code"
4. Status Endpoint
   method: GET
   path: /realms/{realm}/device/status
   @Path: "status"

I would like to hear your opinion.

Michito-Okai commented:
@pedroigor 
I am in the middle of changing the places that were pointed out at the review now.
I am planning to commit day after tomorrow.

pedroigor commented:
> User Code Verification Endpoint's @path overlaps Device Authorization Endpoint's @path. Therefore I want to set them as the following.

They should not clash because they would be producing different media types. The user code verification POST endpoint produces HTML whether the device authorization endpoint is JSON.

JAX-RS takes into account the media type too when resolving the invoker. The changes you mentioned should be available in my branch.

Michito-Okai commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/OAuth2DeviceAuthorizationGrantTest.java:
```diff
@@ -0,0 +1,533 @@
+    }
+
+    @Test
+    public void ConsentCancelTest() throws Exception {
```
I fixed.

Michito-Okai commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/OAuth2DeviceAuthorizationGrantTest.java:
```diff
@@ -0,0 +1,533 @@
+        Assert.assertEquals("slow_down", tokenResponse.getError());
+
+        // Wait the interval
+        WaitUtils.pause(5000);
```
I fixed `testPooling()` and `testDevicePollingIntervalPerClient()` with reference to `LoginTest.loginExpiredCodeWithExplicitRemoveExpired()`. `testDevicePollingIntervalPerClient()` succeeded, but `testPooling()` failed. If I used `WaitUtils.pause`, `testPooling()` succeeded. Is my fix wrong? If it is wrong, could you inform me of the right fix?

mposolda replied:
I replied to this in the comment below: https://github.com/keycloak/keycloak/pull/7731#issuecomment-797375404

Michito-Okai commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/OAuth2DeviceAuthorizationGrantTest.java:
```diff
@@ -0,0 +1,533 @@
+        verificationPage.assertCurrent();
+        verificationPage.submit(response.getUserCode());
+
+        verificationPage.assertLoginPage();
```
I fixed.

mposolda replied:
Thanks!

Michito-Okai commented:
Docs PR https://github.com/keycloak/keycloak-documentation/pull/1134

mposolda commented:
@Michito-Okai @pedroigor Regarding the test, I've added the fix for the test into my branch https://github.com/mposolda/keycloak/commits/KEYCLOAK-7675 . See latest commit from this branch: https://github.com/mposolda/keycloak/commit/296fa00e2de7823d3b8ec5cef44bb7a973eb3d19 . Maybe you can just cherry-pick this commit into your branch and add it to this PR?

Details of the fix: Calling to "setTimeOffset(5)" will convince KeycloakServer to consider the time to be 5 seconds ahead of real time. So if you want to simulate more waits in the test, you need to add the offsets on top of each other. In other words to simulate this:
```
WaitUtils.pause(5);
something1();
WaitUtils.pause(5);
something2();
WaitUtils.pause(10);
something3();
```
you need to do something like this:
```
setTimeOffset(5);
something1();
setTimeOffset(10);
something2();
setTimeOffset(20);
something3();
```
Also I cleanup imports and moved some stuff to the "Try" block to make sure that "finally" executes even if some error happens earlier.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,415 @@
+        return createAuthenticationSession(client, (AuthorizationEndpointRequest) null);
+    }
+
+    public AuthenticationSessionModel createAuthenticationSession(ClientModel client, AuthorizationEndpointRequest request) {
```
Some cleanup will be nice here. This method is never called with the "request" as an argument now. So we can probably just keep single method with the signature `createAuthenticationSession(ClientModel client)` and remove entirely the one with signature `createAuthenticationSession(ClientModel client, AuthorizationEndpointRequest request)` . Also we can remove all the blocks for `if (request != null)` .

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java:
```diff
@@ -0,0 +1,415 @@
+        return authenticationSession;
+    }
+
+    private SessionCodeChecks checksForCode(String authSessionId, String code, String execution, String clientId, String tabId,
```
This method is never used now and can be removed now.

mposolda reviewed (changes requested):
@Michito-Okai @pedroigor Thanks for all the updates! I've resolved most of my comments.

From my point of view, the only remaining things in this PR are:
- Fix the test failure. I've added separate commit for that, which you can cherry-pick (See above)
- We can remove entirely the changes in class ClientModel and OIDCAdvancedConfigWrapper from this PR and move all the configuration to the OAuth2DeviceConfig (See my comment above)
- We can do minor cleanup and remove some dead code in the DeviceEndpoint - see comments above for method `createAuthenticationSession` and `checksForCode` on this class

I think all those changes are easy and we're very close to have this PR merged

mposolda commented:
@Michito-Okai @pedroigor In the end, I've added all the fixes I suggested (Cleanup of DeviceEndpoint unused stuff, Centralize all configuration to the OAuth2DeviceConfig class) in the commit in my branch https://github.com/mposolda/keycloak/commits/KEYCLOAK-7675 . If you cherry-pick the top commit from this branch, we should be fine with this PR IMO (as long as tests will pass, which I hope will be the case)

pedroigor commented:
@Michito-Okai @mposolda Yes we should. By merging the latest changes from yours and my branches we should get it finally merged.

Michito-Okai commented:
@pedroigor  @mposolda 
Thank you for the review and changes.
I cherry-pick the top commit from https://github.com/mposolda/keycloak/commits/KEYCLOAK-7675 . 

I changed the following.
* Time set in L422 and L504 of `OAuth2DeviceAuthorizationGrantTest.java`.
The polling interval is the time between the latest request and the request that return `authorization_pending`. The polling interval is not the time between the latest request and the request that return `slow_down`. Therefore I think time of L422 is 10 and time of L504 is 15.

Could you confirm it?

mposolda reviewed (approved):
@Michito-Okai Thanks! The PR looks good to me now. Approving.

The test failure in unit tests is unrelated to this PR. Leaving to @pedroigor to finally approve and merge this whenever needed.

pedroigor commented:
@mposolda @Michito-Okai Thanks.

@Michito-Okai Thanks for your patience during the review and for promptly working on the changes herein asked. Sometimes we went far beyond a simple review of the functionality itself. Note though that in addition to the device flow functionality, we were able to understand better our limitations when extending Keycloak like you did so that we can improve in the future how similar contributions are done.

Also, we should now focus on CIBA and apply some common patterns from here there.

Without further due, let us merge this :)

dteleguin commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/OAuthClient.java:
```diff
@@ -860,6 +862,71 @@ public AccessTokenResponse doRefreshTokenRequest(String refreshToken, String pas
+                parameters.add(new BasicNameValuePair(OAuth2Constants.SCOPE, scope));
+            }
+            if (nonce != null) {
+                parameters.add(new BasicNameValuePair(OIDCLoginProtocol.NONCE_PARAM, scope));
```
Just noticed this after 3 years :) should be `nonce` not `scope` @wadahiro

pedroigor replied:
@dteleguin Good to know you're around :) Any chance we get a PR from you fixing this?

dteleguin replied:
Better late than never :) https://github.com/keycloak/keycloak/pull/23298
